### PR TITLE
fix broadcast export to onnx

### PR DESCRIPTION
**Pull request by anderspapitto**

### CODE REVIEW

**Review by bddppq — APPROVED** [review]
LG

**Comment by anderspapitto** [comment]
(still WIP, please don't merge yet)

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 1]
```diff
@@ -152,7 +106,8 @@ void fuseBroadcast(std::shared_ptr<Graph>& graph) {
 
     n->replaceInput(input_index, unexpanded_rhs);
     n->i_(attr::broadcast, 1);
-    if (axis) {
+    // Gemm doesn't support the axis argument, so be sure to omit it for that op.
+    if (axis && n->kind() != onnx::Gemm) {
```
Comment (ezyang) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: What if the axis is not the default one?

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 2]
```diff
@@ -52,66 +52,20 @@ bool isBroadcasting(Node* node) {
...
-// return the `axis` argument to be emitted to ONNX on the broadcasting operator
-//
-// Note that this is NOT equivalent to numpy broadcasting semantics, and do
-// not represent the generalized broadcasting that Pytorch implements.
-// Rather, this is Caffe2-style broadcasting.
-//
-// Return value is 1) Whether this expand is fusable, 2) the `axis` argument we
-// should emit to ONNX. Coming from a Pytorch frontend, this should either not
-// be emitted (if we're broadcasting trailing dimensions) or it should be
-// emitted as `0` (leading dimensions.)
```
Comment (ezyang) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: The comment is gone... so what are the semantics now, and why are they this way?

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 3]
```diff
@@ -52,66 +52,20 @@ bool isBroadcasting(Node* node) {
...
 std::tuple<bool, at::optional<size_t>> fusibleExpandTo(at::IntList from, at::IntList to) {
```
Comment (ezyang) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: The tuple seems unnecessary now?

File: torch/csrc/jit/passes/onnx/peephole.cpp (hunk 1 above)
Comment (anderspapitto) [reply torch/csrc/jit/passes/onnx/peephole.cpp]: ++ can enforce that it is

File: torch/csrc/jit/passes/onnx/peephole.cpp (hunk 2 above)
Comment (anderspapitto) [reply torch/csrc/jit/passes/onnx/peephole.cpp]: ++ will add comment

File: torch/csrc/jit/passes/onnx/peephole.cpp (hunk 3 above)
Comment (anderspapitto) [reply torch/csrc/jit/passes/onnx/peephole.cpp]: yes, an optional size is enough information now

**Comment by ezyang** [comment]
```
17:39:22 torch/csrc/jit/passes/onnx/peephole.cpp:113:24: error: 'Gemm' was not declared in this scope
17:39:22        if (n->kind() == Gemm) {
17:39:22                         ^
```

**Comment by anderspapitto** [comment]
oops :)

**Review by ezyang — APPROVED** [review]

**Comment by anderspapitto** [comment]
the onnx-fb-universe failure is a spurious CI git issue

**Pull request merged**